integrations/access/accessrequest/app.go

```diff
@@ -0,0 +1,447 @@
+			return PluginData{}, trace.CompareFailed("existing sentMessages is empty")
+		}
+
+		count := len(reqReviews)
+		oldCount = existing.ReviewsCount
+		if oldCount >= count {
+			return PluginData{}, trace.AlreadyExists("reviews are sent already")
+		}
+
+		existing.ReviewsCount = count
+		return existing, nil
+	})
+	if trace.IsAlreadyExists(err) {
+		logger.Get(ctx).Debug("Failed to post reply: replies are already sent")
+		return nil
+	}
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	slice := reqReviews[oldCount:]
+	if len(slice) == 0 {
+		return nil
+	}
+
+	errors := make([]error, 0, len(slice))
+	for _, data := range pd.SentMessages {
+		ctx, _ = logger.WithFields(ctx, logger.Fields{"channel_id": data.ChannelID, "message_id": data.MessageID})
+		for _, review := range slice {
+			if err := a.bot.PostReviewReply(ctx, data.ChannelID, data.MessageID, review); err != nil {
+				errors = append(errors, err)
+			}
+		}
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// getMessageRecipients takes an access request and returns a list of channelIDs that should be messaged.
+// channelIDs can represent any communication channel depending on the MessagingBot implementation:
+// a public channel, a private one, or a user direct message channel.
+func (a *App) getMessageRecipients(ctx context.Context, req types.AccessRequest) []common.Recipient {
+	log := logger.Get(ctx)
+
+	// We receive a set from GetRawRecipientsFor but we still might end up with duplicate channel names.
+	// This can happen if this set contains the channel `C` and the email for channel `C`.
+	recipientSet := common.NewRecipientSet()
+
+	switch a.pluginType {
+	case types.PluginTypeServiceNow:
+		// The ServiceNow plugin does not use recipients currently and create incidents in the incident table directly.
+		// Recipients just needs to be non empty.
+		recipientSet.Add(common.Recipient{})
+		return recipientSet.ToSlice()
+	case types.PluginTypeOpsgenie:
+		if recipients, ok := req.GetSystemAnnotations()[types.TeleportNamespace+types.ReqAnnotationSchedulesLabel]; ok {
+			for _, recipient := range recipients {
+				rec, err := a.bot.FetchRecipient(ctx, recipient)
+				if err != nil {
+					log.Warning(err)
+				}
+				recipientSet.Add(*rec)
+			}
+			return recipientSet.ToSlice()
+		}
+	}
+
+	validEmailSuggReviewers := []string{}
+	for _, reviewer := range req.GetSuggestedReviewers() {
+		if !lib.IsEmail(reviewer) {
+			log.Warningf("Failed to notify a suggested reviewer: %q does not look like a valid email", reviewer)
+			continue
+		}
+
+		validEmailSuggReviewers = append(validEmailSuggReviewers, reviewer)
+	}
+	rawRecipients := a.recipients.GetRawRecipientsFor(req.GetRoles(), validEmailSuggReviewers)
+	for _, rawRecipient := range rawRecipients {
+		recipient, err := a.bot.FetchRecipient(ctx, rawRecipient)
+		if err != nil {
+			// Something wrong happened, we log the error and continue to treat valid rawRecipients
+			log.Warning(err)
```

Comment:
I'd add more context to the log here so it's clear what we were trying to do when we encountered an error